txnsync/transactionCache.go

```diff
@@ -169,31 +198,100 @@ func (lru *transactionCache) acknowledge(seqs []uint64) {
 
 // add a given transaction ID to the short term cache.
 func (st *shortTermTransactionCache) add(txid transactions.Txid) {
-	if st.transactionsMap[txid] {
+	entry, exists := st.transactionsMap[txid]
+	if exists {
+		// promote
+		if entry.detach() {
+			// there are other elements on the list.
+			// if the given entry happen to be the first entry, then pick
+			// the next entry.
+			if entry == st.first {
+				st.first = entry.next
+			}
+			// add to the tail of the list.
+			entry.addToList(st.first)
+		} else {
+			// no other elements on the list -
+			// nothing to do in this case.
+		}
 		return
 	}
+
 	mapLen := len(st.transactionsMap)
 	if mapLen >= st.size {
 		// we reached size, delete the oldest entry.
-		delete(st.transactionsMap, st.transactionsIDs[st.oldest])
-		st.transactionsIDs[st.oldest] = txid
-		st.transactionsMap[txid] = true
-		st.oldest = (st.oldest + 1) % st.size
+		t := st.first
+
+		// disconnect the current one; no need to test return code since we know
+		// there will be more elements on the list.
+		t.detach()
+
+		// replace the first entry with the next one.
+		st.first = t.next
+
+		// delete the current value from the map.
+		delete(st.transactionsMap, t.txid)
+
+		// copy the new transaction id into the existing object.
+		copy(t.txid[:], txid[:])
+
+		// place the new entry as the last entry on the list.
+		t.addToList(st.first)
+
+		// add the new entry to the map
+		st.transactionsMap[txid] = t
 		return
 	}
-	st.transactionsIDs[mapLen] = txid
-	st.transactionsMap[txid] = true
+
+	// grab an entry from the free list ( if any )
+	entry = st.free
+	if entry != nil {
+		if entry.detach() {
+			st.free = entry.next
+		} else {
+			st.free = nil
+		}
+		copy(entry.txid[:], txid[:])
+	} else {
+		// the free list doesn't have an entry - allocate a new one.
+		entry = &shortTermCacheEntry{
+			txid: txid,
+		}
+	}
+	if st.first == nil {
+		st.first = entry
+		entry.next = entry
+		entry.prev = entry
+	} else {
+		entry.addToList(st.first)
+	}
+	st.transactionsMap[txid] = entry
 }
 
 // contained checks if the given transaction id presents in the short term cache
 func (st *shortTermTransactionCache) contained(txid transactions.Txid) bool {
-	return st.transactionsMap[txid]
+	return st.transactionsMap[txid] != nil
 }
 
 // reset clears the short term cache
 func (st *shortTermTransactionCache) reset() {
-	st.transactionsMap = make(map[transactions.Txid]bool, st.size)
-	st.oldest = 0
+	if st.first == nil {
+		return
+	}
+	st.transactionsMap = make(map[transactions.Txid]*shortTermCacheEntry, st.size)
+	if st.free == nil {
+		st.free = st.first
+		st.first = nil
+		return
+	}
+	used := st.first
+	free := st.free
+	free.prev.next = used
+	used.prev.next = free
+	lastFree := free.prev
+	free.prev = used.prev
+	used.prev = lastFree
+	st.first = nil
```

Comment:
could you please add a test that calls `reset` in 2 used and 3 free nodes and ensures `st.free` has all 5 nodes reachable in both directions?

Reply:
sure.